### PR TITLE
#8357 - System should not load monomer to the library if modification…

### LINKED ISSUE #8357: System should not load monomer to the library if modificationTypes value contain inproper symbols
**Steps to Reproduce**
1. Open Macromolecules - Flex mode (clean canvas)
2. Go to console and execute following commands:

<details><summary>_Peptide1 with tab in modificationTypes value</summary>
<p>

```javascript
await ketcher.updateMonomersLibrary('\n  -INDIGO-10122515372D\n\n  0  0  0  0  0  0  0  0  0  0  0 V3000\nM  V30 BEGIN CTAB\nM  V30 COUNTS 1 0 0 0 0\nM  V30 BEGIN ATOM\nM  V30 1 _Peptide1 11.975 -7.85 0.0 0 CLASS=AA SEQID=1\nM  V30 END ATOM\nM  V30 BEGIN BOND\nM  V30 END BOND\nM  V30 END CTAB\nM  V30 BEGIN TEMPLATE\nM  V30 TEMPLATE 1 AA/_Peptide1/_Peptide1/ NATREPLACE=AA/A\nM  V30 BEGIN CTAB\nM  V30 COUNTS 4 3 3 0 0\nM  V30 BEGIN ATOM\nM  V30 1 C -0.75 0.0 0.0 0\nM  V30 2 C 0.25 0.0 0.0 0\nM  V30 3 H 0.75 0.866 0.0 0\nM  V30 4 H 0.75 -0.866 0.0 0\nM  V30 END ATOM\nM  V30 BEGIN BOND\nM  V30 1 2 1 2\nM  V30 2 1 2 3\nM  V30 3 1 2 4\nM  V30 END BOND\nM  V30 BEGIN SGROUP\nM  V30 1 SUP 1 ATOMS=(1 3) XBONDS=(1 2) BRKXYZ=(9 -0.250000 -0.433000 0.00000-\nM  V30 0 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000) LABEL=H CLASS-\nM  V30 =LGRP\nM  V30 2 SUP 2 ATOMS=(1 4) XBONDS=(1 3) BRKXYZ=(9 -0.250000 0.433000 0.000000-\nM  V30  0.000000 0.000000 0.000000 0.000000 0.000000 0.000000) LABEL=H CLASS=-\nM  V30 LGRP\nM  V30 3 SUP 3 ATOMS=(2 1 2) XBONDS=(2 2 3) BRKXYZ=(9 0.250000 0.433000 0.000-\nM  V30 000 0.250000 -0.433000 0.000000 0.000000 0.000000 0.000000) LABEL=_Pep-\nM  V30 tide1 CLASS=AA SAP=(3 2 3 Al) SAP=(3 2 4 Br) NATREPLACE=AA/A\nM  V30 END SGROUP\nM  V30 END CTAB\nM  V30 END TEMPLATE\nM  END\n>  <type>\nmonomerTemplate\n\n>  <modificationTypes>\n \t ;\n\n$$$$\n', { format: 'sdf' })
``` 

</p>
</details> 

**Actual behavior**
`_Peptide1` amino acid appears in the library
Image

**Expected behavior**
Monomer doesn't appear at the Library
System throws an error in console: `Load of "_Peptide1" monomer has failed, monomer definition contains invalid modificationTypes value. The modificationTypes couldn't be empty` 

As per [requirement](https://github.com/epam/Indigo/issues/3161):

>   - `modificationTypes` - string, optional. Used for aminoacids modification through the context menu.
Format: Any symbol except formatting ones, max length 200, spaces allowed
If value doesn't fit format - throw an error

**Environment details:**
- Ketcher Version 3.9.0-rc.1 Build at 2025-10-02; 16:35:40
- Indigo Version 1.37.0-rc.1.0-gca09660df-wasm32-wasm-clang-19.0.0
- Chrome Version 141.0.7390.55 (Official Build) (64-bit)
- Win10

Related issue: https://github.com/epam/Indigo/issues/3161

### PR DESCRIPTION
**Pull request by PaulinaKacz**
Fix #8357: Validate modificationTypes field in monomer library loading                                                                                                                                                          
                                       
  🐛 Original Issue                                                                                                                                                                                                               
                                       
  GitHub Issue: #8357
  Title: System should not load monomer to the library if modificationTypes value contains improper symbols

  Reproduction:
  // Open Macromolecules - Flex mode
  // Execute in console:
  await ketcher.updateMonomersLibrary('...SDF with modificationTypes: \t ;...', { format: 'sdf' })

  Expected behavior:
  - Monomer should NOT appear in the library
  - System should throw error: Load of "_Peptide1" monomer has failed, monomer definition contains invalid modificationTypes value. The modificationTypes couldn't be empty

  Actual behavior:
  - ❌ Monomer with invalid modificationTypes appeared in the library

  📋 Requirements

  According to specification:
  modificationTypes - string, optional. Used for amino acids modification through the context menu.
  Format: Any symbol except formatting ones, max length 200, spaces allowed
  If value doesn't fit format: throw an error

  Invalid formats:
  - Empty arrays: []
  - Whitespace-only: [" ", "\t", "\n"]
  - Formatting characters only: ["\t ;"]
  - Exceeding 200 characters total length

  ✅ Solution

  1. Added validation function (packages/ketcher-core/src/utilities/monomers.ts)
  - isValidModificationTypes() - validates format and content
  - MODIFICATION_TYPES_MAX_LENGTH = 200 constant
  - MODIFICATION_TYPES_EMPTY_ERROR_MESSAGE constant

  Validation rules:
  - undefined/not provided → valid (optional field)
  - Empty array [] → invalid
  - Whitespace-only content → invalid
  - Total length > 200 chars → invalid
  - Any non-whitespace content → valid

  2. Integrated validation in library loading (packages/ketcher-core/src/application/editor/Editor.ts)
  - Validation runs during updateMonomersLibrary()
  - Invalid monomers are rejected and added to skippedItems
  - Error logged via KetcherLogger.error()

  3. Added comprehensive unit tests (packages/ketcher-core/__tests__/utilities/monomers.test.ts)
  - 41 test cases covering all validation scenarios
  - Edge cases: empty arrays, whitespace, formatting chars, length limits

  4. Fixed type definition (packages/ketcher-react/src/script/editor/Editor.ts)
  - Made modificationTypes optional in SaveNewMonomerData type
  - Prevents passing empty arrays instead of undefined

  5. Updated monomer creation wizard (MonomerCreationWizard.tsx)
  - Pass undefined instead of empty array when no modification types selected
  - Prevents false validation errors for non-Peptide monomers

  🧪 Testing

  Unit tests:
  - ✅ 41 tests in monomers.test.ts - all passing
  - ✅ Updated tests in Editor.test.ts

  E2E tests:
  - ✅ All 57 previously failing autotests now pass
  - ✅ Tests covering Base, Sugar, Phosphate, Nucleotide, CHEM monomer creation
  - ✅ Tests for preset creation and library updates

  Manual testing:
  - ✅ Invalid modificationTypes rejected during library loading
  - ✅ Valid monomers created successfully
  - ✅ Error messages displayed correctly

  📝 Files Changed

  Core validation logic:
  - packages/ketcher-core/src/utilities/monomers.ts - validation function
  - packages/ketcher-core/src/application/editor/Editor.ts - integration
  - packages/ketcher-core/__tests__/utilities/monomers.test.ts - unit tests
  - packages/ketcher-core/__tests__/application/editor/Editor.test.ts - integration tests

  Type fixes:
  - packages/ketcher-react/src/script/editor/Editor.ts - optional type
  - packages/ketcher-react/src/script/ui/views/components/MonomerCreationWizard/MonomerCreationWizard.tsx - pass undefined

  🎯 Impact

  Security: Prevents malformed monomers from being loaded into the library
  Data integrity: Ensures modificationTypes field follows specification
  User experience: Clear error messages when invalid monomers are rejected
  Backward compatibility: Existing valid monomers continue to work (undefined is valid)


## Check list
- [x] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [x] reviewers are notified about the pull request

### CODE REVIEW

File: packages/ketcher-core/src/utilities/monomers.ts [hunk 1]
```diff
@@ -131,3 +136,46 @@ export function isMonomerSgroupWithAttachmentPoints(monomer: BaseMonomer) {
...
+
+  // Check if all elements contain only whitespace/formatting characters
+  // \s matches all whitespace characters including tabs, newlines, carriage returns, etc.
+  const hasNonWhitespaceContent = modificationTypes.some((type) => {
+    const trimmed = type.replace(/\s+/g, '');
+    return trimmed.length > 0;
+  });
+
+  return hasNonWhitespaceContent;
+}
```
Comment (AlexeyGirin) [inline packages/ketcher-core/src/utilities/monomers.ts:180]: Please trim spaces before validations

**Review by AlexeyGirin — DISMISSED** [review]

File: packages/ketcher-core/src/utilities/monomers.ts (hunk 1 above)
Comment (PaulinaKacz) [reply packages/ketcher-core/src/utilities/monomers.ts:180]: Fixed

**Review by svvald — APPROVED** [review]
